src/orange/orangeinp/CsgTypes.cc

```diff
@@ -79,6 +81,85 @@ std::string to_string(Node const& n)
     return os.str();
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from bottom/top z segments.
+ */
+SpecialTrapezoid::SpecialTrapezoid(ZSegment&& bot, ZSegment&& top)
+    : bot_(std::move(bot)), top_(std::move(top))
+{
+    constexpr auto left = Bound::lo;
+    constexpr auto right = Bound::hi;
+
+    CELER_VALIDATE(bot_.z < top_.z,
+                   << "bottom segment must have a lower z than the top "
+                      "segment");
+
+    // Calculate abs_tol_ based on extents
+    auto r_min = std::min(bot_.r[left], top_.r[left]);
+    auto r_max = std::max(bot_.r[right], top_.r[right]);
+    Real3 const extents{r_max - r_min, top_.z - bot_.z, 0};
+    abs_tol_ = ::celeritas::detail::BumpCalculator(
+        Tolerance<>::from_default())(extents);
+
+    // Determine variety
+    SoftClose soft_close(abs_tol_);
+
+    bool has_pointy_bot = soft_close(bot_.r[left], bot_.r[right]);
+    bool has_pointy_top = soft_close(top_.r[left], top_.r[right]);
+
+    CELER_VALIDATE(!(has_pointy_bot && has_pointy_top),
+                   << "must contain at least 3 distinct points");
+
+    if (has_pointy_bot)
+    {
+        CELER_VALIDATE(top_.r[left] < top_.r[right],
+                       << "r values must appear in increasing order");
+        variety_ = Variety::pointy_bot;
+    }
+    else if (has_pointy_top)
+    {
+        CELER_VALIDATE(bot_.r[left] < bot_.r[right],
+                       << "r values must appear in increasing order");
+        variety_ = Variety::pointy_top;
+    }
+    else
+    {
+        CELER_VALIDATE(
+            top_.r[left] < top_.r[right] && bot_.r[left] < bot_.r[right],
+            << "r values must appear in increasing order");
+        variety_ = Variety::quad;
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the unique points in clockwise order, starting with the bottom right.
+ */
+auto SpecialTrapezoid::unique_points() const -> VecReal2
+{
+    constexpr auto left = Bound::lo;
+    constexpr auto right = Bound::hi;
+
+    VecReal2 points;
+
+    // Add bottom points
+    points.push_back({bot_.r[right], bot_.z});
+    if (variety_ != Variety::pointy_bot)
+    {
+        points.push_back({bot_.r[left], bot_.z});
+    }
+
+    // Add top points
+    points.push_back({top_.r[left], top_.z});
+    if (variety_ != Variety::pointy_top)
+    {
+        points.push_back({top_.r[right], top_.z});
+    }
+
```

Comment:
Having an extra `variety_` class and that logic really seems like overengineering; if you rely on point deduplication upstream you would just have:
```suggestion
    points.push_back({bot_.r[right], bot_.z});
    if (bot_.r[right] != bot_.r[left])
    {
        points.push_back({bot_.r[left], bot_.z});
    }
    points.push_back({top_.r[left], top_.z});
    if (top_.r[right] != top_.r[left])
    {
        points.push_back({top_.r[right], top_.z});
    }
```
which seems just as clear.

Also, not a big deal, but do we have a "default" ordering for polygon points? For the extruded prism and other shapes, we move counterclockwise from the top right quadrant (i.e. theta = 0, theta increasing) rather than bottom right clockwise (theta = -pi/2, theta decreasing).

Reply:
The reason I do point deduplication on the class is to prevent construction in an invalid state, i.e. two sets of equal points. There will be a second class, `SpecialTrapezoidDecomposer`, that will actually construct these objects, so having robust DBC for construction will lessen the DBC burden within `SpecialTrapezoidDecomposer`.

Reply:
For arbitrary polygons, e.g., `ExtrudedPolygon` we use clockwise, same a Geant4. I was thinking that this function would return such a polygon. However, it's easy enough to change.

Reply:
Let's add a to-do in the code consistency post to change these to counterclockwise later: https://github.com/celeritas-project/celeritas/issues/721 so we can do it at the same time as the existing poly